File: src/MiddleManServer.cpp
#include "MiddleManServer.h"
#include <iostream>
using namespace std;

Server::Server()
    : context(1), socket(context, zmq::socket_type::rep)
{
    socket.bind("tcp://*:5555");
}

void Server::run()
{
    while (true)
    {
        zmq::message_t request;
        socket.recv(request);
        string cmd = string(static_cast<char *>(request.data()), request.size());

        if (cmd == "GET")
        {
            string data = "gk";
            zmq::message_t reply(data.size());
            memcpy(reply.data(), data.c_str(), data.size());
            socket.send(reply, zmq::send_flags::none);
        }
        else
        {
            // Send a reply to acknowledge the setField command
            zmq::message_t reply(2); // "OK" as an acknowledgment
            memcpy(reply.data(), "OK", 2);
            socket.send(reply, zmq::send_flags::none);
        }
    }
}

File: java-maven-project/src/main/java/org/example/MiddleManServer.java
package org.example;

import org.zeromq.ZMQ;

public class MiddleManServer {
    private final ZMQ.Socket socket;

    public MiddleManServer(int port) {
        ZMQ.Context context = ZMQ.context(1);
        this.socket = context.socket(ZMQ.ROUTER);
        this.socket.bind("tcp://*:" + port);
    }

    public void run() {

        while (!Thread.currentThread().isInterrupted()) {


            System.out.println("Server Running");
            // Receive client request
            byte[] clientId = socket.recv(0);
            byte[] request = socket.recv(0);

            String Nrequest = new String(request, ZMQ.CHARSET);
            System.out.println("Request: "+Nrequest);

            // Simulate processing
            String response = "Response from Load Balancer";

            // Reply to the client
            socket.sendMore(clientId);
            socket.send(response.getBytes(ZMQ.CHARSET), 0);
        }
    }


    //For independant execution
    public static void main(String[] args) {
        // Example usage:
        MiddleManServer loadBalancer = new MiddleManServer(5555);
        loadBalancer.run();
    }
}
